### DISCUSSION TITLE
Plugin dev and logging

### DISCUSSION THREAD

**Discussion by dainok** [netbox-community/netbox#10418] [Q&A] [answered]
Hello,

I'm developing a complex plugin and I'm stuck on a error: it happens when I load a specific page. Seems that the Python stacktrace is hidden, and I get only a NoReverseMatch:

~~~
<class 'django.urls.exceptions.NoReverseMatch'>

'netdoc' is not a registered namespace

Python version: 3.8.10
NetBox version: 3.3.2-dev
~~~

Any chance to get the full stacktrace from a plugin error?
Thank you

**Comment by candlerb** [comment]
Set EMAIL and ADMINS in configuration.py.  The backtrace should be sent by E-mail to the ADMINS.  Alternatively, if you set DEBUG=True in configuration.py, you should get an interactive backtrace in your browser.

(This works for normal Netbox, but I've not checked either option with a broken plugin)

**Comment by dainok** [comment]
Hello @candlerb thank you for your answer. I turned DEBUG on, and now I can see the stack trace in the web page, but it's limited to the "is not a registered namespace" exception. I also enabled EMAIL and ADMIN, but I can see no email (I'm using localhost which I manually tested and it works).

**Reply by candlerb** [reply]
It's ADMINS not ADMIN.

**Reply by dainok** [reply]
Sorry, I wrote ADMIN but I configured ADMINS:

~~~
ADMINS = [
    ('Andrea', 'root'),
]
~~~

**Reply by candlerb** [reply]
Check your MTA logs to see what happened to it.

**Comment by dainok** [comment]
There is no connection to the MTA. tcpdump shows no packages to port 25.

**Reply by candlerb** [reply]
OK, then maybe the exception is happening before the code which catches it and turns it into an E-mail.  (What happens if you put a `raise RuntimeError("foo")` inside some code that normally works?)

The error message appears to be coming from Django innards. Some possible clues here:
https://stackoverflow.com/questions/41883254/django-is-not-a-registered-namespace

Otherwise sorry, I can't help any more than that.

**Comment by dainok** [comment]
Same result (I mean no TCP connection on port 25).
I added the command  raise RuntimeError("foo") to a view that is normally working.
In the web page I can read 'netdoc' is not a registered namespace inside 'plugins'. If I run "makemigrations", I can read RuntimeError: foo.

**Reply by candlerb** [reply]
> In the web page I can read 'netdoc' is not a registered namespace inside 'plugins'.

Yeah, I expect you'll have to fix that first - or temporarily remove the page or view that's causing it.

**Reply by dainok** [reply]
That's my problem: I have an error on that page, I cannot debug it because netbox unregisters the plugin removing the stacktrace. I'm stuck on that.

**Reply by candlerb** [reply]
The last-ditch approach (which unfortunately I have had to use recently) is a binary chop:
* comment out half your code
* see if the problem recurs
* then you know which half is bad, so comment out half of *that* code
* repeat until you find the offending line(s)

If you could find a way to invoke a plugin outside of Netbox, that might also work.  For some errors, simply "import"ing it might be enough.

**Reply by dainok** [reply]
Thank you for your support. It could not be so easy, speaking in general. Most of the time I found a quick and dirty approach: invoking makemigrations help me to see the buggy line.

But in this specific case, I encounter the error while I'm clicking the "delete selected" after flagging "select all" in a specific form. I'm sure I'm missing something, but I cannot figure out which one. I'm referring to this one: https://github.com/dainok/netdoc/issues/6
I have the same error on all forms in my plugin.

**Reply by candlerb** [reply]
What's the actual URL shown in the browser URL bar when this error appears?

My guess is it's when you do a multi-select delete, you get a confirmation page listing the things about to be deleted - and possibly it's trying to turn them into hyperlinks, and is failing.  Like this (`/tenancy/contacts/delete/`):

![image](https://user-images.githubusercontent.com/44789/191460912-f648109d-ade0-422f-92a2-65a7954f72ed.png)

In this table, the contact name and group name are hyperlinks.

If you can reproduce this with a minimal plugin that does nothing else, then it might be a bug in the Netbox plugin framework.

**Reply by candlerb** [reply]
Incidentally, why are you using a non-release branch of Netbox? You should be using 'master', which will give you v3.3.4 currently.

There is no support for upgrading between unreleased versions of Netbox, and any attempts to do so may result in your database getting into an inconsistent state.

**Reply by dainok** [reply]
The URL is correct (http://localhost:8000/plugins/netdoc/credentials/delete/) and also the post is. I'm surely missing something in my code.

I'm trying to figure out how the plugin is loaded. Can  you point me in the right direction? I see something in extras/plugin.

**Reply by candlerb** [reply]
Sorry, I don't know the internals of plugin loading.

If my theory is correct, the error occurs while rendering `netbox/templates/generic/bulk_delete.html`. You could try removing this section:

```
    <div class="table-responsive">
      {% render_table table 'inc/table.html' %}
    </div>
```

and see if the problem goes away. If so, that at least gives you a clue as to what's going on, and if you wish you can investigate further into `netbox/templates/inc/table.html`

I'm afraid I don't know if this problem is with Netbox itself not correctly supporting bulk deletions from plugins, or with the way your plugin is written.  As I say, if you can boil this down to the absolute minimal plugin that demonstrates the issue, then I expect one of the developers can make a call.

**Reply by dainok** [reply]
Thank you for the suggestion and for the support. I'm working on it

**Comment by dainok** [comment, marked as answer]
I found the line causing the issue: default_return_url

~~~
class CredentialBulkDeleteView(generic.BulkDeleteView):
    queryset = models.Credential.objects.all()
    table = tables.CredentialTable
    default_return_url = 'netdoc:credential_list'
    filterset = filtersets.CredentialFilterSet
~~~

The I missed "plugins", the right line is:

~~~
    default_return_url = 'plugins:netdoc:credential_list'
~~~

For some reason netbox is hiding the exceptions so it wasn't easy to find what was going on.

**Reply by jeremystretch** [reply]
> For some reason netbox is hiding the exceptions so it wasn't easy to find what was going on.

As @candlerb points out above, NetBox isn't hiding anything. The exception is reported per the error message, and stack traces are emailed (or you can enable debugging).

**Reply by dainok** [reply]
Thank you @jeremystretch. My mistake, I was referring to my netbox setup not netbox itself. As you can read from the thread, I wasn't able to get stacktrace via browser or via email. Again, it probably happens because of my setup, but I didn't get why.
